Add automation to update App Store strings .po file for Jetpack

mokagio opened the pull request:
Adds rough logic to the `update_appstore_strings` lane to update `.po` file for Jetpack as well as WordPress.

cc @AliSoftware. I'm not sure where you're at with the Android side of this, but I hope this will be useful, even though for the Play Store we use the `PlayStoreStrings.po`.

## To test

1. Checkout **a new branch** from this one. `update_appstore_strings` commits and pushes the changes it makes, being on a new branch will prevent it from adding noise to this PR.
1. Optional: Push the branch. If you don't, `update_appstore_strings` will successfully update the `.po` but will fail because it can't push the changes 🤷‍♂️
2. Edit some of the files in `fastlane/appstoreres/jetpack_metadata/`
3. Run `bundle exec fastlane update_appstore_strings version:X`, where `X` is a version value.

You should see a new commit in your history. Looking at it you should see your changes reflected in the `WordPress/Jetpack/Resources/AppStoreStrings.po` file, including the "what's new" version key changing to the version value `X` you specified.

☝️ In your test you might notice a new line in the description, e.g. for WordPress

```diff
 #. translators: Subtitle to be displayed below the application name in the Apple App Store. Limit to 30 characters including spaces and commas!
 msgctxt "app_store_subtitle"
-msgid "Build a website or blog"
+msgid "Build a website or blog
+"
```

That's some kind of bug which I haven't research yet, but is unrelated to these changes.

## Regression Notes
1. Potential unintended areas of impact
N.A.: This does not update any existing automation, it only adds new steps.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
N.A.

3. What automated tests I added (or what prevented me from doing so)
N.A.

---

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes. **N.A.**
- [x] I have considered adding accessibility improvements for my changes. **N.A.**
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary. **N.A.**

mokagio committed:
Track Jetpack App Store strings sources for localization

mokagio committed:
Add logic to update Jetpack .po to update_appstore_strings lane

mokagio committed:
Add AppStoreStrings.po file for Jetpack

mokagio commented on WordPress/Jetpack/Resources/AppStoreStrings.po:
```diff
@@ -0,0 +1,109 @@
+"Site stats in your\n"
+"pocket.\n"
+msgstr ""
+
```
The release toolkit adds this new line, which really bothers me... I'm not sure why that happens here and not in the other file. 

Any ideas?

mokagio commented on WordPress/Jetpack/Resources/AppStoreStrings.po:
```diff
@@ -0,0 +1,109 @@
+# Translation of Release Notes & Apple Store Description in English (US)
```
To generated this file:

1. I copied the WordPress one
2. I deleted the irrelevant `msgctxt`+`msgid`+`msgstr` groups
3. I ensured all the `msgid` values matched what's currently on GlotPress

I then run the lane and verified that the original `msgstr` values from WordPress were replaced with the content of the Jetpack `.txt` files as configured by the lane.

mokagio commented on WordPress/Jetpack/Resources/release_notes.txt:
```diff
@@ -0,0 +1,15 @@
+
+Last up, we squashed a bug that sometimes caused the app to crash during comment moderation and another to fix a login error for WordPress.org sites to show inline.
+
+That’s all for 17.6. See you in 17.7!
```
These are the same notes as WordPress, with the exception of the paragraph about the Reader, because that's not available in Jetpack.

mokagio commented on WordPress/Resources/AppStoreStrings.po:
```diff
@@ -40,7 +41,7 @@ msgctxt "app_store_keywords"
 msgstr ""
 
-msgctxt "v17.6-whats-new"
+msgctxt "v17.9-whats-new"
```
@mokagio fixup this debug code!

mokagio commented on fastlane/appstoreres/jetpack_metadata/source/app_name.txt:
```diff
@@ -0,0 +1 @@
+Jetpack: WP Security & Speed
```
Unless otherwise specified, I copied the `.txt` content from what was already on GlotPress.

---

A note on this file. As mentioned above, this is a value that's in GlotPress, but I couldn't find a matching `.txt` file for the WordPress localizations 🤔

mokagio commented on fastlane/appstoreres/jetpack_metadata/source/subtitle.txt:
```diff
@@ -0,0 +1 @@
+Make your WordPress site faster and safer.
```
I came up with this because it's required by release toolkit, and it's anyway a good idea to have it.

I'm sure there's something better we can use, but maybe we'll save it for the next release? This is please is better than having no subtitle, which is the current state.

<img width="657" alt="Screen Shot 2021-06-18 at 4 39 48 pm" src="https://user-images.githubusercontent.com/1218433/122517833-d63b3700-d053-11eb-90a8-2ba8c22ff5ad.png">

@leandroalonso please advise 😄 🙇

leandroalonso replied:
What's the char limitation of this field? I think we can use `Get powerful security and performance tools in your pocket.` which is the JPAndroid subtitle. :)

AliSoftware replied:
@leandroalonso Apparently 30, according to the number at the bottom right of this textfield. Which means your suggestion doesn't fit and is 29 chars too long 😢 

![image](https://user-images.githubusercontent.com/216089/122807977-8915ca00-d2cc-11eb-8073-988267489d89.png)

mokagio commented on fastlane/Fastfile:
```diff
@@ -179,6 +198,44 @@ platform :ios do
+      source_files: files,
+      release_version: options[:version]
+    )
+  end
```
There's admittedly some duplication between this and the WordPress lane. I think we can live with that for the moment and think about removing it later if needed.

jkmassel replied:
Agreed – copy/paste is our friend here – no point over-abstracting

mokagio commented on fastlane/Fastfile:
```diff
@@ -179,6 +198,44 @@ platform :ios do
+  #####################################################################################
+  desc 'Updates the AppStoreStrings.po file for the WordPress app with the latest data'
+  lane :update_jetpack_appstore_strings do |options|
+    source_metadata_folder = File.join(PROJECT_ROOT_FOLDER, 'fastlane', 'appstoreres', 'jetpack_metadata', 'source')
```
I chose this path to stay in line with https://github.com/wordpress-mobile/WordPress-iOS/pull/16571.

momo-ozawa commented on fastlane/appstoreres/jetpack_metadata/source/promo_screenshot_5.txt:
```diff
@@ -0,0 +1,2 @@
+Back Up your site
```
This should be "Back up" -- lowercase u 😄

mokagio replied:
```suggestion
Back up your site
```

momo-ozawa commented on fastlane/Fastfile:
```diff
@@ -179,6 +198,44 @@ platform :ios do
+  #####################################################################################
+  # update_jetpack_appstore_strings
+  # -----------------------------------------------------------------------------------
+  # This lane updates the AppStoreStrings.po files for the WordPress app with
```
[nit]
```suggestion
  # This lane updates the AppStoreStrings.po files for the Jetpack app with
```

mokagio replied:
ZOMG! Too much _copy pasta_!

momo-ozawa commented on fastlane/Fastfile:
```diff
@@ -179,6 +198,44 @@ platform :ios do
+  # Example:
+  # bundle exec fastlane update_jetpack_appstore_strings version:10.7
+  #####################################################################################
+  desc 'Updates the AppStoreStrings.po file for the WordPress app with the latest data'
```
[nit]
```suggestion
  desc 'Updates the AppStoreStrings.po file for the Jetpack app with the latest data'
```

jkmassel replied:
Nit – there's a `Jetpack-Fastfile` that this method should maybe go in?

The idea was to keep everything grouped by app (even if there are call-sites in this file for JP)

mokagio replied:
Gotcha! Will move it there, then and address @momo-ozawa's docs fixes suggestions.

mokagio replied:
Done in 94480f8f3 🙇

mokagio commented on fastlane/appstoreres/jetpack_metadata/source/keywords.txt:
```diff
@@ -0,0 +1 @@
+social,notes,jetpack,writing,geotagging,media,blog,website,blogging,journal
```
I'm pretty sure these were copied from the WordPress iOS app, right @leandroalonso @momo-ozawa ?

I'm not well versed in App Store Optimization, but I think it's redundant to include the app name in the keywords. Might be worth checking and updating for the next release.

momo-ozawa replied:
Yeah, looks like it was copied from the WPiOS app.

> I'm not well versed in App Store Optimization, but I think it's redundant to include the app name in the keywords. Might be worth checking and updating for the next release.

Just want to point out that our app name is "Jetpack: WP Security and Speed" -- not sure if that means we'd want to keep the "jetpack" keyword or not.

jkmassel replied:
The keywords weren't specified, so were copied. IMHO those are all still applicable, though we could probably focus in more on JP-specific things. Also IMHO "jetpack" is an important keyword to have because we don't have the "Jetpack" name all by itself

momo-ozawa reviewed (commented):
@mokagio 👋 
If I run the `update_appstore_strings` lane, the `Wordpress/Jetpack/Resources/AppStoreStrings.po` is NOT updated. Only the `Wordpress/Resources/AppStoreStrings.po` is updated. Strange, since I can see that the `update_appstore_strings` lane calls both the wordpress and jetpack lanes

AliSoftware commented on WordPress/Jetpack/Resources/AppStoreStrings.po:
```diff
@@ -0,0 +1,109 @@
+
+#. translators: Subtitle to be displayed below the application name in the Apple App Store. Limit to 30 characters including spaces and commas!
+msgctxt "app_store_subtitle"
+msgid "Make your WordPress site faster and safer.
+"
```
I think this added newline is a known bug in release-toolkit that either @JavonDavis or @jkmassel worked on?

(We might want to `.chomp` the text read from the `.txt` to get rid of newlines here I guess)

jkmassel replied:
Yep that'd be the fix

AliSoftware reviewed (approved):
Tested using the following steps:
 - Created a dummy branch `test/pr16706`, and pushed it to remote
 - Edited `WordPress/Jetpack/Resources/release_notes.txt` to add some dummy text to the notes
 - Same with `WordPress/Resources/release_notes.txt`
 - Commited both changes to end up with a non-dirty working copy
 - Ran `fl update_appstore_strings version:123.45`
 - Looked at `git diff jetpack-app-store-localization-automation` to check what changed, which seems correct (see below)

> In your test you might notice a new line in the description, e.g. for WordPress

I confirm that I encountered this issue (see diff below), and I think it's a known one, unrelated to your changes but that we definitively need to fix.

<details><summary><tt>git diff jetpack-app-store-localization-automation</tt></summary>

```diff
diff --git a/WordPress/Jetpack/Resources/AppStoreStrings.po b/WordPress/Jetpack/Resources/AppStoreStrings.po
index e199260eb8..71a64442af 100644
--- a/WordPress/Jetpack/Resources/AppStoreStrings.po
+++ b/WordPress/Jetpack/Resources/AppStoreStrings.po
@@ -39,8 +39,10 @@ msgctxt "app_store_keywords"
 msgid "social,notes,jetpack,writing,geotagging,media,blog,website,blogging,journal"
 msgstr ""
 
-msgctxt "v17.6-whats-new"
+msgctxt "v123.45-whats-new"
 msgid ""
+"This is Jetpack!!\n"
+"\n"
 "With 17.6, you can add the Audio Block to any WordPress.com site, whatever plan you’re on – including the Free Plan. The Audio Block makes it easy to add podcasts, music, and sound files to your posts and pages.\n"
 "\n"
 "We’ve been hard at work wrangling UX improvements, too:\n"
diff --git a/WordPress/Jetpack/Resources/release_notes.txt b/WordPress/Jetpack/Resources/release_notes.txt
index ffd542975c..7690606bae 100644
--- a/WordPress/Jetpack/Resources/release_notes.txt
+++ b/WordPress/Jetpack/Resources/release_notes.txt
@@ -1,3 +1,5 @@
+This is Jetpack!!
+
 With 17.6, you can add the Audio Block to any WordPress.com site, whatever plan you’re on – including the Free Plan. The Audio Block makes it easy to add podcasts, music, and sound files to your posts and pages.
 
 We’ve been hard at work wrangling UX improvements, too:
diff --git a/WordPress/Resources/AppStoreStrings.po b/WordPress/Resources/AppStoreStrings.po
index 56a25d934f..8b5dbce69e 100644
--- a/WordPress/Resources/AppStoreStrings.po
+++ b/WordPress/Resources/AppStoreStrings.po
@@ -15,7 +15,8 @@ msgstr ""
 
 #. translators: Subtitle to be displayed below the application name in the Apple App Store. Limit to 30 characters including spaces and commas!
 msgctxt "app_store_subtitle"
-msgid "Build a website or blog"
+msgid "Build a website or blog
+"
 msgstr ""
 
 #. translators: Multi-paragraph text used to display in the Apple App Store.
@@ -40,8 +41,10 @@ msgctxt "app_store_keywords"
 msgid "social,notes,jetpack,writing,geotagging,media,blog,website,blogging,journal"
 msgstr ""
 
-msgctxt "v17.6-whats-new"
+msgctxt "v123.45-whats-new"
 msgid ""
+"This is WordPress :)\n"
+"\n"
 "With 17.6, you can add the Audio Block to any WordPress.com site, whatever plan you’re on – including the Free Plan. The Audio Block makes it easy to add podcasts, music, and sound files to your posts and pages.\n"
 "\n"
 "In the Reader, you’ll now see a count and summary of Likes added to every post. If that’s not enough Likes in your life, tap the summary and you’ll be able to see the full list.\n"
diff --git a/WordPress/Resources/release_notes.txt b/WordPress/Resources/release_notes.txt
index 2de20c1100..ddadf32339 100644
--- a/WordPress/Resources/release_notes.txt
+++ b/WordPress/Resources/release_notes.txt
@@ -1,3 +1,5 @@
+This is WordPress :)
+
 With 17.6, you can add the Audio Block to any WordPress.com site, whatever plan you’re on – including the Free Plan. The Audio Block makes it easy to add podcasts, music, and sound files to your posts and pages.
 
 In the Reader, you’ll now see a count and summary of Likes added to every post. If that’s not enough Likes in your life, tap the summary and you’ll be able to see the full list.
```
</details>

AliSoftware commented:
> cc @AliSoftware. I'm not sure where you're at with the Android side of this, but I hope this will be useful, even though for the Play Store we use the PlayStoreStrings.po.

Since @JavonDavis is handling updating automation and Fastfile for Jetpack, I think he's the one who might be more interested in this – I'll let that to him to avoid stepping on each other's toes.

Indeed for Android we use `PlayStoreStrings.po` and not `AppStoreStrings.po` but that's relatively similar; the main difference is that we don't use automation to auto-uploading/downloading the metadata from the PlayStore (like we do with `fastlane deliver` for iOS).

AliSoftware commented:
> @mokagio 👋
> If I run the `update_appstore_strings` lane, the `Wordpress/Jetpack/Resources/AppStoreStrings.po` is NOT updated. Only the `Wordpress/Resources/AppStoreStrings.po` is updated. Strange, since I can see that the `update_appstore_strings` lane calls both the wordpress and jetpack lanes

@momo-ozawa:
 - Did you change some source strings like in `WordPress/Jetpack/Resources/release_notes.txt` and `WordPress/Resources/release_notes.txt` before testing the lane? Because since Gio already ran the lane on this branch, running it again yourself without any change should lead to no change at all – but in practice because of [this bug](https://github.com/wordpress-mobile/WordPress-iOS/pull/16706#discussion_r655253396), leads to the WordPress one showing a change with that extra newline.
 - Another reason why the Jetpack one was not updated, is if you didn't push your test branch before trying the lane. Currently the lane first updates the `.po` file for WordPress, them commits, then update the po file for JP, then commits (as opposed to a single commit for both). so if it failed to push the commit for the WordPress file (because your test branch was not pushed/tracked by remote) it would crash after the update to WP but before having a chance to update the JP file.

jkmassel reviewed (approved):
Left one concern, but then ok from my side pending everyone else's feedback

mokagio committed:
Fix copy for Jetpack 5th screenshot

Thanks @momo-ozawa

mokagio committed:
Move Jetpack App Store lane to Jetpack-Fastfile

Thanks @jkmassel

See
https://github.com/wordpress-mobile/WordPress-iOS/pull/16706#discussion_r655612689
